validation.go

```diff
@@ -222,46 +226,61 @@ func validateParamValue(key string, value string, param Parameter) error {
 	return multiErr
 }
 
-// assignParamDefaults fills any empty configuration with its assigned default value
+// initConfig checks for unrecognized params, and fills any empty configuration with its assigned default value
 // returns an error if a parameter is not recognized.
-func assignParamDefaults(params map[string]Parameter, config map[string]string) (map[string]string, error) {
+func (v validator) initConfig(config map[string]string) (map[string]string, error) {
 	output := make(map[string]string)
 	var multiErr error
 	for key, val := range config {
-		if _, ok := params[key]; !ok {
+		// trim key and val
+		key = strings.TrimSpace(key)
+		val = strings.TrimSpace(val)
+		// check for unrecognized params
+		if _, ok := v[key]; !ok {
 			multiErr = multierr.Append(multiErr, fmt.Errorf("unrecognized parameter %q", key))
 			continue
 		}
+		// set config map and assign defaults
 		output[key] = val
-		if len(strings.TrimSpace(val)) == 0 {
-			output[key] = params[key].Default
+		if val == "" {
+			output[key] = v[key].Default
+		}
+	}
+	// set defaults for unassigned params
+	for key, param := range v {
+		if _, ok := output[key]; !ok {
+			output[key] = param.Default
 		}
 	}
 	return output, multiErr
 }
 
 // validateParamType validates that a parameter value is parsable to its assigned type.
-func validateParamType(p Parameter, key string, value string) error {
-	switch p.Type {
+func (v validator) validateParamType(key string, value string) error {
+	// empty value is valid for all types
```

Comment:
Good catch! 👍

Reply:
not really a catch.. it was checked in the function calling this one.. I just moved it 👀